### DISCUSSION TITLE
Dashboard errors when managing users

### DISCUSSION THREAD

**Discussion by TheOtherBrian1** [supabase/supabase#21247] [Troubleshooting]
_This is a copy of a troubleshooting article on Supabase's docs site. It may be missing some details from the original. View the [original article](https://supabase.com/docs/guides/troubleshooting/dashboard-errors-when-managing-users-N1ls4A)._

These error are normal a side effect of issues in your custom user management logic. This can cause errors that return HTTP `500` status codes with a of code `unexpected_failure` and one of the following error messages

*   Failed to create user: Database error creating new user
*   Failed to update user: Error updating user
*   Failed to delete user: Database error deleting user
*   Database error updating user
*   Database error saving new user

## Debugging this error

*   [Auth logs](https://supabase.com/dashboard/project/_/logs/auth-logs): as the error relates to [Auth](https://supabase.com/docs/guides/auth) users
*   [Postgres logs](https://supabase.com/dashboard/project/_/logs/postgres-logs): for raw error logs related to database

## Common causes of this error:

*   Trigger/trigger function setup on the `auth.users` table
*   A constraint on the `auth.users` table which isn't being met
*   You are using Prisma and it has broken all the permissions on the `auth.users` table

## Example error messages

Use the hints provided in the error message to fix issues in your custom user management logic.

**Trigger/trigger function related error messages** - [Solution for trigger related issues](https://supabase.com/docs/guides/troubleshooting/dashboard-errors-when-managing-users-N1ls4A#solution-for-trigger-related-issues)

    "error":"error update user`s last_sign_in field: ERROR: permission denied for table profiles (SQLSTATE 42501)"

**Constraint related error message** - [Solution for constraint related issues](https://supabase.com/docs/guides/troubleshooting/dashboard-errors-when-managing-users-N1ls4A#solution-for-constraint-related-issues)

    ERROR:  23503: update or delete on table "users" violates foreign key constraint "profiles_id_fkey" on table "profiles"
    DETAIL:  Key (id)=(7428a53c-75b7-4531-9ae9-1567d9c4ac0a) is still referenced from table "profiles".

**Missing column**

    ERROR: column \"updated_at\" of relation \"profiles\" does not exist (SQLSTATE 42703)

**Broken search path / incorrect name** - [42P01 related solution](https://supabase.com/docs/guides/troubleshooting/resolving-42p01-relation-does-not-exist-error-W4_9-V)

    failed to close prepared statement: ERROR: current transaction is aborted, commands ignored until end of transaction block (SQLSTATE 25P02): ERROR: relation \"public.profiles\" does not exist (SQLSTATE 42P01)

***

## Solution for constraint related issues

*   Check for foreign/primary key relationship between the `auth.users` table and another table
*   Then ALTER the [behavior](https://stackoverflow.com/questions/5383612/setting-up-table-relations-what-do-cascade-set-null-and-restrict-do) of the relationship and recreate it with a less [restrictive constraint](https://stackoverflow.com/questions/3359329/how-to-change-the-foreign-key-referential-action-behavior).
*   If this is related to deleting records, review the [Cascade Deletes doc](https://supabase.com/docs/guides/database/postgres/cascade-deletes) for possible approaches (e.g. using `CASCADE` / `SET NULL`)

***

## Solution for trigger related issues

Supabase Auth uses your project's database to store user data. It relies on the `auth` schema, and Supabase restricts access to the `auth` schema to prevent unintended custom changes that could break the functionality of the Auth service.

Check if the `auth` schema contains any triggers in the [Dashboard's trigger section](https://supabase.com/dashboard/project/_/database/triggers?schema=auth).

1.  Identify related functions using `security invoker` from the [Dashboard's function section](https://supabase.com/dashboard/project/_/database/functions)
2.  Remove all triggers by dropping their functions with a CASCADE modifier shown below (This command still works because the `postgres` role has the ownership of the function, and the `CASCADE` clause will drop the trigger indirectly.)

```sql
DROP FUNCTION <function name>() CASCADE;

-- If you'd prefer, you can drop the trigger alone with the following query:
-- DROP TRIGGER <trigger_name> on auth.<table_name>;
```

3.  Recreate the functions with a [security definer](https://supabase.com/docs/guides/database/functions#security-definer-vs-invoker) modifier
4.  Recreate the triggers

**Example function and trigger using security definer**
The [SQL Editor](https://supabase.com/dashboard/project/_/sql/) contains a template for [User Management](https://supabase.com/dashboard/project/_/sql/examples). Within it, there is a working example of how to setup triggers with security definer that may be worth referencing:

```sql
create table profiles (
  id uuid references auth.users on delete cascade not null primary key,
  updated_at timestamp with time zone,
  username text unique,
  full_name text,
  avatar_url text,
  website text,

  constraint username_length check (char_length(username) >= 3)
);

create function public.handle_new_user()
returns trigger
set search_path = ''
as $$
begin
  insert into public.profiles (id, full_name, avatar_url)
  values (new.id, new.raw_user_meta_data->>'full_name', new.raw_user_meta_data->>'avatar_url');
  return new;
end;
$$ language plpgsql security definer;

create trigger on_auth_user_created
  after insert on auth.users
  for each row execute procedure public.handle_new_user();
```

### Explanation

One of the most common design patterns in Supabase is to add a trigger to the `auth.users` table. The database role managing authentication (`supabase_auth_admin`) only has the necessary permissions it needs to perform its duties. So, when a trigger operated by the `supabase_auth_admin` interacts outside the auth schema, it causes a permission error.

A security definer function retains the privileges of the database user that created it. As long as it is the `postgres` role, your auth triggers should be able to engage with outside tables.

**Comment by MattSolvesGames** [comment]
`new.raw_app_meta_data  ->> 'custom_id'` isn't returning anything except null values for me even though I can see on the `auth.users` row that the user has the metadata set.

```
CREATE TABLE users (
    id UUID REFERENCES auth.users ON DELETE CASCADE NOT NULL PRIMARY KEY,
    custom_id VARCHAR(17)
);

CREATE FUNCTION handle_new_user() RETURNS TRIGGER LANGUAGE plpgsql SECURITY definer SET search_path = '' AS $$
BEGIN
    INSERT INTO public.users(id, custom_id) VALUES (new.id, new.raw_app_meta_data ->> 'custom_id');
    RETURN new;
END;
$$;

CREATE TRIGGER on_auth_user_created
    AFTER INSERT ON auth.users
    FOR EACH ROW
    EXECUTE FUNCTION handle_new_user();
```

Here's what I see on the `auth.users` entry:
```
  "raw_app_meta_data": {
    "provider": "email",
    "custom_id": "73849099371620888",
    "providers": [
      "email"
    ]
  },
```

Yet on the new row created in `public.users`, the column is null, which was giving me a 500 error since I had specified a `NOT NULL` constraint. As far as I can tell, I'm doing everything exactly like the documentation and other discussions have said. Is the metadata not getting created with the rest of the row or something?

**Reply by TheOtherBrian1** [reply]
I don't believe custom_id is a default value on the `raw_app_meta_data` column. You can add it with an [auth_hook](https://supabase.com/docs/guides/auth/auth-hooks?queryGroups=language&language=add-admin-role).

Maybe you were thinking of the [`raw_user_meta_data` column, which can be more easily modified through a Supabase-Client](https://supabase.com/docs/reference/javascript/auth-signup)?

@MattSolvesGames For your case, _just for the debugging period_, I recommend placing a `RAISE LOG` keyword in your function. It will allow you to inspect the values in the trigger to ensure they meet your needs:
```sql
-- The REPLACE keyword overwrites the previous function if it already exists
CREATE OR REPLACE FUNCTION handle_new_user() 
RETURNS TRIGGER 
LANGUAGE plpgsql 
SECURITY definer SET search_path = '' 
AS $$
BEGIN
    -- record new row to Postgres Logs | Remove when you're done debugging
    RAISE LOG 'logged handle_new_user: %', (select to_jsonb(new.*));

    INSERT INTO public.users(id, custom_id) 
    VALUES (new.id, new.raw_app_meta_data ->> 'custom_id');

    RETURN new;
END;
$$;
```

Once triggered, you can then inspect the new row value in the [Postgres Logs](https://supabase.com/dashboard/project/_/logs/postgres-logs). Just make sure to filter by the term "logged handle_new_user":
<img width="909" alt="Screenshot 2024-05-21 at 10 01 18 PM" src="https://github.com/supabase/supabase/assets/91111415/f78c920c-0195-490f-b876-6b930926934e">

Alternatively, you can filter the advanced Log Explorer for errors from the `supabase_auth_admin` role ([guide here](https://github.com/orgs/supabase/discussions/26224)). The log manager digests logs in batches, so it may take a minute after executing the function before it can be searched.

In either case, the output isn't going to be the most readable, so you should run it through a [JSON Beautifier.](https://jsonformatter.org/)

**Reply by MattSolvesGames** [reply]
@TheOtherBrian1 Huh; there was more to my comment than got posted; I guess my edit didn't go through--the purpose was supposed to be to mention to others finding this thread that there's currently a bug where `raw_app_meta_data` is actually added to the row _after_  it's created, unlike with `raw_user_meta_data`. Therefore, if you're trying to pass app metadata (like my `custom_id` property) added during creation to a trigger, it won't work and you'll always get `null` back. It appears this is being tracked [here](https://github.com/supabase/auth/issues/1280); I just wanted to mention it here for others since it's more likely they'll stumble on this thread than that one if they aren't specifically searching for it. Thanks for the helpful debugging tips for the future though ♥

**Reply by rungchan2** [reply]
thank you everyone! this helped me a lot! 😄
